### PR TITLE
Port lxc to the new client library

### PR DESCRIPTION
**Pull request by stgraber**
This converts all of the lxc/* commands to use the new client package.
It then removes the experimental flag on the new package and adds a deprecation note to the old one.

Going through this work, I've noticed a lot of opportunities for refactoring inside our client, but I think this is best left for another time (probably next cycle as we look at reworking the client tool).

The next step will be adding support to CopyContainer for choosing between "pull", "push" and "relay" copy modes. The library code will deal with the actual negotiation and relaying code that's needed for that.

### CODE REVIEW

File: test/suites/basic.sh [hunk 1]
```diff
@@ -52,8 +52,8 @@ test_basic_usage() {
 
   # Test custom filename for image export
   lxc image export testimage "${LXD_DIR}/foo"
-  [ "${sum}" = "$(sha256sum "${LXD_DIR}/foo.tar.xz" | cut -d' ' -f1)" ]
-  rm "${LXD_DIR}/foo.tar.xz"
+  [ "${sum}" = "$(sha256sum "${LXD_DIR}/foo" | cut -d' ' -f1)" ]
+  rm "${LXD_DIR}/foo"
```
Comment (stgraber) [inline test/suites/basic.sh]: I wonder what's that doing in this commit, will move around.

File: test/suites/remote.sh [hunk 2]
```diff
@@ -17,7 +17,6 @@ test_remote_url() {
 
   for url in ${urls}; do
     lxc_remote remote add test "${url}"
-    lxc_remote finger test:
```
Comment (stgraber) [inline test/suites/remote.sh]: Can't remember the reason for this change, will investigate.

Comment (stgraber) [inline test/suites/remote.sh]: It's because you can't finger (GetServer()) a non-LXD server. This was accidentally working with the old code.

**Comment by stgraber** [comment]
Tracked down the ETag failure and fixed my own comments, hopefully Jenkins will be happy now.

**Comment by stgraber** [comment]
So looks like there's a bug somewhere in the background op tracking code. That's manifesting as two things in the testsuite, both are near impossible to reproduce locally:
 - Missing output in "lxc exec"
 - Hung commands (mostly "lxc delete")

File: test/suites/basic.sh (hunk 1 above)
Comment (stgraber) [inline test/suites/basic.sh]: This got moved.

**Comment by stgraber** [comment]
There's clearly still an issue with "lxc exec"...

**Comment by stgraber** [comment]
I'm unable to reproduce the exec issue locally, so I made a best guess as to what it may be, lets see if that's it.

**Comment by stgraber** [comment]
Yay, jenkins is finally happy.

**Comment by stgraber** [comment]
And another clean run. History since I fixed the two main issues:
 - Pass
 - Fail (due to ZFS ebusy)
 - Pass

So looks like things are pretty solid now.

**Comment by brauner** [comment]
Scheduled for Sunday. Wohoooo, client API. :)

File: lxc/info.go [hunk 3]
```diff
@@ -38,16 +39,23 @@ func (c *infoCmd) flags() {
...
 
-func (c *infoCmd) run(config *lxd.Config, args []string) error {
+func (c *infoCmd) run(conf *config.Config, args []string) error {
 	var remote string
 	var cName string
+	var err error
 	if len(args) == 1 {
-		remote, cName = config.ParseRemoteAndContainer(args[0])
+		remote, cName, err = conf.ParseRemote(args[0])
+		if err != nil {
```
Comment (brauner) [inline lxc/info.go]: You likely wanted to have the error handling after the cases, right? :) Just an fyi.

File: lxc/init.go [hunk 4]
```diff
@@ -144,22 +145,31 @@ func (c *initCmd) flags() {
...
+	if err != nil {
+		return err
+	}
 
 	var name string
 	var remote string
 	if len(args) == 2 {
-		remote, name = config.ParseRemoteAndContainer(args[1])
+		remote, name, err = conf.ParseRemote(args[1])
+		if err != nil {
```
Comment (brauner) [inline lxc/init.go]: same here :)

File: lxc/launch.go [hunk 5]
```diff
@@ -37,22 +38,31 @@ func (c *launchCmd) flags() {
...
+	if err != nil {
+		return err
+	}
 
 	var name string
 	var remote string
 	if len(args) == 2 {
-		remote, name = config.ParseRemoteAndContainer(args[1])
+		remote, name, err = conf.ParseRemote(args[1])
+		if err != nil {
```
Comment (brauner) [inline lxc/launch.go]: here too

File: lxc/publish.go [hunk 6]
```diff
@@ -50,12 +49,22 @@ func (c *publishCmd) run(config *lxd.Config, args []string) error {
...
+	cRemote, cName, err := conf.ParseRemote(args[0])
+	if err != nil {
+		return err
+	}
+
 	if len(args) >= 2 && !strings.Contains(args[1], "=") {
 		firstprop = 2
-		iRemote, iName = config.ParseRemoteAndContainer(args[1])
+		iRemote, iName, err = conf.ParseRemote(args[1])
+		if err != nil {
```
Comment (brauner) [inline lxc/publish.go]: error handling after the cases? :)

File: lxc/exec.go [hunk 7]
```diff
@@ -214,6 +237,6 @@ func (c *execCmd) run(conf *config.Config, args []string) error {
 		termios.Restore(cfd, oldttystate)
 	}
 
-	os.Exit(ret)
-	return fmt.Errorf(i18n.G("unreachable return reached"))
+	os.Exit(int(op.Metadata["return"].(float64)))
```
Comment (brauner) [inline lxc/exec.go]: a `float64`?

File: lxc/file.go [hunk 8]
```diff
@@ -69,6 +70,145 @@ func (c *fileCmd) flags() {
...
 
+func (c *fileCmd) recursivePullFile(d lxd.ContainerServer, container string, p string, targetDir string) error {
+	buf, resp, err := d.GetContainerFile(container, p)
+	if err != nil {
+		return err
+	}
+
+	target := filepath.Join(targetDir, filepath.Base(p))
+	if resp.Type == "directory" {
+		if err := os.Mkdir(target, os.FileMode(resp.Mode)); err != nil {
```
Comment (brauner) [inline lxc/file.go]:
```
if err := foo(); err != nil {
}
```
shouldn't that be

```
err := foo()
if err != nil {
}
```

:)

File: lxc/init.go (hunk 4 above)
Comment (stgraber) [inline lxc/init.go]: Yeah, similarly intentional :)

File: lxc/publish.go (hunk 6 above)
Comment (stgraber) [inline lxc/publish.go]: Same :)

File: lxc/exec.go (hunk 7 above)
Comment (stgraber) [inline lxc/exec.go]: Yeah, that's what JSON ints are unpacked as, trying to parse as anything else causes a panic.

File: lxc/file.go (hunk 8 above)
Comment (stgraber) [inline lxc/file.go]: It's code I moved from client.go and my OCD didn't seem to trigger on that one, fixing :)

**Comment by stgraber** [comment]
Pushed a rebased copy of the tree, this only fixes the "if err := ..." stuff in lxc/file.go

File: lxc/exec.go (hunk 7 above)
Comment (brauner) [inline lxc/exec.go]: Weird. :)

**Pull request merged**